File: PE-diagram/src/nl/adaptivity/process/clientProcessModel/ClientProcessModel.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.clientProcessModel;

import net.devrieze.util.security.SimplePrincipal;
import nl.adaptivity.diagram.Bounded;
import nl.adaptivity.diagram.Rectangle;
import nl.adaptivity.process.diagram.DiagramNode;
import nl.adaptivity.process.diagram.LayoutAlgorithm;
import nl.adaptivity.process.processModel.EndNode;
import nl.adaptivity.process.processModel.ProcessModelBase;
import nl.adaptivity.process.processModel.ProcessNodeSet;
import nl.adaptivity.process.processModel.engine.IProcessModelRef;
import nl.adaptivity.process.util.Identifiable;

import java.util.*;


public abstract class ClientProcessModel<T extends ClientProcessNode<T, M>, M extends ClientProcessModel<T,M>> extends ProcessModelBase<T, M> {

  public static final String NS_JBI = "http://adaptivity.nl/ProcessEngine/activity";

  public static final String NS_UMH = "http://adaptivity.nl/userMessageHandler";

  public static final String NS_PM = "http://adaptivity.nl/ProcessEngine/";

  static final String PROCESSMODEL_NS = NS_PM;

  private double mTopPadding = 5d;
  private double mLeftPadding = 5d;
  private double mBottomPadding = 5d;
  private double mRightPadding = 5d;

  LayoutAlgorithm<T> mLayoutAlgorithm;

  private boolean mNeedsLayout = false;

  public ClientProcessModel(UUID uuid, final String name, final Collection<? extends T> nodes) {
    this(uuid, name, nodes, new LayoutAlgorithm<T>());
  }

  public ClientProcessModel(UUID uuid, final String name, final Collection<? extends T> nodes, LayoutAlgorithm<T> layoutAlgorithm) {
    super(nodes);
    setName(name);
    mLayoutAlgorithm = layoutAlgorithm == null ? new LayoutAlgorithm<T>() : layoutAlgorithm;
    setNodes(nodes);
    setUuid(uuid == null ? UUID.randomUUID() : uuid);
  }

  protected ClientProcessModel() {
    this(null, null, new ArrayList<T>(), null);
  }

  public abstract T asNode(final Identifiable id);

  public void setNodes(final Collection<? extends T> nodes) {
    super.setModelNodes(ProcessNodeSet.processNodeSet(nodes));
    invalidate();
  }

  public LayoutAlgorithm<T> getLayoutAlgorithm() {
    return mLayoutAlgorithm;
  }

  public void setLayoutAlgorithm(LayoutAlgorithm<T> layoutAlgorithm) {
    mLayoutAlgorithm = layoutAlgorithm;
  }

  public double getVertSeparation() {
    return mLayoutAlgorithm.getVertSeparation();
  }


  public void setVertSeparation(double vertSeparation) {
    if (mLayoutAlgorithm.getVertSeparation()!=vertSeparation) {
      invalidate();
    }
    mLayoutAlgorithm.setVertSeparation(vertSeparation);
  }


  public double getHorizSeparation() {
    return mLayoutAlgorithm.getHorizSeparation();
  }


  public void setHorizSeparation(double horizSeparation) {
    if (mLayoutAlgorithm.getHorizSeparation()!=horizSeparation) {
      invalidate();
    }
    mLayoutAlgorithm.setHorizSeparation(horizSeparation);
  }

  public double getDefaultNodeWidth() {
    return mLayoutAlgorithm.getDefaultNodeWidth();
  }


  public void setDefaultNodeWidth(double defaultNodeWidth) {
    if (mLayoutAlgorithm.getDefaultNodeWidth()!=defaultNodeWidth) {
      invalidate();
    }
    mLayoutAlgorithm.setDefaultNodeWidth(defaultNodeWidth);
  }


  public double getDefaultNodeHeight() {
    return mLayoutAlgorithm.getDefaultNodeHeight();
  }


  public void setDefaultNodeHeight(double defaultNodeHeight) {
    if (mLayoutAlgorithm.getDefaultNodeHeight()!=defaultNodeHeight) {
      invalidate();
    }
    mLayoutAlgorithm.setDefaultNodeHeight(defaultNodeHeight);
  }


  public double getTopPadding() {
    return mTopPadding;
  }


  public void setTopPadding(double topPadding) {
    double offset = topPadding-mTopPadding;
    for(T n: getModelNodes()) {
      n.setY(n.getY()+offset);
    }
    mTopPadding = topPadding;
  }


  public double getLeftPadding() {
    return mLeftPadding;
  }


  public void setLeftPadding(double leftPadding) {
    double offset = leftPadding-mLeftPadding;
    for(T n: getModelNodes()) {
      n.setX(n.getX()+offset);
    }
    mLeftPadding = leftPadding;
  }


  public double getBottomPadding() {
    return mBottomPadding;
  }


  public void setBottomPadding(double bottomPadding) {
    mBottomPadding = bottomPadding;
  }


  public double getRightPadding() {
    return mRightPadding;
  }


  public void setRightPadding(double rightPadding) {
    mRightPadding = rightPadding;
  }


  public void invalidate() {
    mNeedsLayout  = true;
  }

  public void resetLayout() {
    for (T n: getModelNodes()) {
      n.setX(Double.NaN);
      n.setY(Double.NaN);
    }
    invalidate();
  }

  public boolean isInvalid() {
    return mNeedsLayout;
  }

  public int getEndNodeCount() {
    int i=0;
    for(T node: getModelNodes()) {
      if (node instanceof EndNode) { ++i; }
    }
    return i;
  }

  @Override
  public IProcessModelRef<T, M> getRef() {
    throw new UnsupportedOperationException("Not implemented");
  }

  public T getNode(String nodeId) {
    for(T n: getModelNodes()) {
      if (nodeId.equals(n.getId())) {
        return n;
      }
    }
    return null;
  }

  public void setOwner(String owner) {
    setOwner(new SimplePrincipal(owner));
  }

  public Collection<? extends ClientStartNode<? extends T, M>> getStartNodes() {
    List<ClientStartNode<? extends T, M>> result = new ArrayList<>();
    for(T n:getModelNodes()) {
      if (n instanceof ClientStartNode) {
        result.add((ClientStartNode<? extends T, M>) n);
      }
    }
    return result;
  }

  public boolean addNode(T node) {
    if (super.addNode(node)) {
      node.setOwnerModel(asM());
      // Make sure that children can know of the change.
      notifyNodeChanged(node);
      return true;
    }
    return false;
  }

  public T removeNode(int nodePos) {
    T node = super.removeNode(nodePos);
    disconnectNode(node);
    return node;
  }

  @Override
  public boolean removeNode(T node) {
    if (node==null) {
      return false;
    }
    if (getModelNodes().remove(node)) {
      disconnectNode(node);
      return true;
    }
    return false;
  }

  private void disconnectNode(T node) {
    node.setPredecessors(Collections.<Identifiable>emptyList());
    node.setSuccessors(Collections.<Identifiable>emptyList());
    notifyNodeChanged(node);
  }

  public void layout() {
    final List<DiagramNode<T>> diagramNodes = toDiagramNodes(getModelNodes());
    if(mLayoutAlgorithm.layout(diagramNodes)) {
      double maxX = Double.MIN_VALUE;
      double maxY = Double.MIN_VALUE;
      for(DiagramNode<T> n:diagramNodes) {
        n.getTarget().setX(n.getX()+getLeftPadding());
        n.getTarget().setY(n.getY()+getTopPadding());
        maxX = Math.max(n.getRight(), maxX);
        maxY = Math.max(n.getBottom(), maxY);
      }
    }
  }

  private List<DiagramNode<T>> toDiagramNodes(Collection<? extends T> modelNodes) {
    HashMap<T,DiagramNode<T>> map = new HashMap<>();
    List<DiagramNode<T>> result = new ArrayList<>();
    for(T node:modelNodes) {
      final double leftExtend;
      final double rightExtend;
      final double topExtend;
      final double bottomExtend;
      if (node instanceof Bounded) {
        boolean tempCoords = Double.isNaN(node.getX())||Double.isNaN(node.getY());
        double tmpX=0;
        double tmpY=0;
        if (tempCoords) {
          tmpX=node.getX();node.setX(0);
          tmpY=node.getY();node.setY(0);
          mNeedsLayout=true; // we need layout as we have undefined coordinates.
        }
        Rectangle bounds = ((Bounded)node).getBounds();
        leftExtend = node.getX()-bounds.left;
        rightExtend = bounds.right()-node.getX();
        topExtend = node.getY()-bounds.top;
        bottomExtend = bounds.bottom()-node.getY();
        if (tempCoords) {
          node.setX(tmpX);
          node.setY(tmpY);
        }
      } else {
        leftExtend = rightExtend = mLayoutAlgorithm.getDefaultNodeWidth()/2;
        topExtend = bottomExtend = mLayoutAlgorithm.getDefaultNodeHeight()/2;
      }
      DiagramNode<T> dn = new DiagramNode<>(node, leftExtend, rightExtend, topExtend, bottomExtend);
      if (node.getId()!=null) {
        map.put(node, dn);
      }
      result.add(dn);
    }

    for(DiagramNode<T> dn:result) {
      T mn = dn.getTarget();
      for(Identifiable successor:mn.getSuccessors()) {
        DiagramNode<T> rightdn = map.get(successor);
        if (rightdn!=null) {
          dn.getRightNodes().add(rightdn);
        }
      }
      for(Identifiable predecessorId:mn.getPredecessors()) {
        T predecessor = getNode(predecessorId);
        DiagramNode<T> leftdn = map.get(predecessor);
        if (leftdn!=null) {
          dn.getLeftNodes().add(leftdn);
        }
      }
    }
    return result;
  }

}

File: PE-diagram/src/main/kotlin/nl.adaptivity.process/clientProcessModel/ClientProcessModel.kt
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.clientProcessModel

import net.devrieze.util.Handle
import net.devrieze.util.Handles
import net.devrieze.util.security.SecurityProvider
import net.devrieze.util.security.SimplePrincipal
import nl.adaptivity.diagram.Bounded
import nl.adaptivity.process.diagram.DiagramNode
import nl.adaptivity.process.diagram.DrawableProcessModel
import nl.adaptivity.process.diagram.DrawableProcessNode
import nl.adaptivity.process.diagram.LayoutAlgorithm
import nl.adaptivity.process.processModel.*
import nl.adaptivity.process.processModel.engine.IProcessModelRef
import nl.adaptivity.process.util.Identifiable
import nl.adaptivity.process.util.Identified
import nl.adaptivity.process.util.Identifier
import nl.adaptivity.process.util.IdentifyableSet
import java.security.Principal
import java.util.*

interface ClientProcessModel : ProcessModel<DrawableProcessNode, DrawableProcessModel?> {
  interface Builder : ProcessModel.Builder<DrawableProcessNode, DrawableProcessModel?> {

  }

  override val rootModel: RootClientProcessModel?

  var layoutAlgorithm: LayoutAlgorithm<DrawableProcessNode>

  fun layout()

}

abstract class RootClientProcessModel : RootProcessModelBase<DrawableProcessNode, DrawableProcessModel?>, MutableRootProcessModel<DrawableProcessNode, DrawableProcessModel?> {

  abstract class Builder : RootProcessModelBase.Builder<DrawableProcessNode, DrawableProcessModel?>, ClientProcessModel.Builder {
    var  layoutAlgorithm: LayoutAlgorithm<DrawableProcessNode>

    constructor(): this(nodes= mutableSetOf())

    constructor(
      nodes: MutableSet<ProcessNode.Builder<DrawableProcessNode, DrawableProcessModel?>> = mutableSetOf(),
      childModels: MutableSet<ChildProcessModel.Builder<DrawableProcessNode, DrawableProcessModel?>> = mutableSetOf(),
      name: String? = null,
      handle: Long = -1L,
      owner: Principal = SecurityProvider.SYSTEMPRINCIPAL,
      roles: MutableList<String> = mutableListOf<String>(),
      uuid: UUID? = null,
      imports: MutableList<IXmlResultType> = mutableListOf<IXmlResultType>(),
      exports: MutableList<IXmlDefineType> = mutableListOf<IXmlDefineType>(),
      layoutAlgorithm: LayoutAlgorithm<DrawableProcessNode> = LayoutAlgorithm()) : super(nodes, childModels, name, handle, owner, roles, uuid, imports, exports) {
      this.layoutAlgorithm = layoutAlgorithm
    }

    constructor(base: RootProcessModel<*,*>) : super(base) {
      this.layoutAlgorithm = (base as? ClientProcessModel)?.layoutAlgorithm ?: LayoutAlgorithm<DrawableProcessNode>()
    }

    abstract override fun build(pedantic: Boolean): RootProcessModelBase<DrawableProcessNode, DrawableProcessModel?>
  }

  override val rootModel: RootClientProcessModel get() = this

  var layoutAlgorithm: LayoutAlgorithm<DrawableProcessNode>

  @JvmOverloads constructor(uuid: UUID? = null, name: String? = null, nodes: Collection<DrawableProcessNode> = emptyList(), layoutAlgorithm: LayoutAlgorithm<DrawableProcessNode> = LayoutAlgorithm<DrawableProcessNode>(), nodeFactory: NodeFactory<DrawableProcessNode, DrawableProcessModel?>) :
    super(nodes, uuid = uuid ?: UUID.randomUUID(), name = name, nodeFactory = nodeFactory) {
    this.layoutAlgorithm = layoutAlgorithm
  }

  @JvmOverloads
  constructor(builder: RootProcessModelBase.Builder<DrawableProcessNode, DrawableProcessModel?>, nodeFactory: NodeFactory<DrawableProcessNode, DrawableProcessModel?>, pedantic: Boolean = builder.defaultPedantic) : super(builder, nodeFactory, pedantic) {
    this.layoutAlgorithm = (builder as? Builder)?.layoutAlgorithm ?: LayoutAlgorithm()
  }

  var topPadding = 5.0
    set(topPadding) {
      val offset = topPadding - this.topPadding
      for (n in modelNodes) {
        n.setY(n.y + offset)
      }
      field = topPadding
    }

  var leftPadding = 5.0
    set(leftPadding) {
      val offset = leftPadding - this.leftPadding
      for (n in modelNodes) {
        n.setX(n.x + offset)
      }
      field = leftPadding
    }

  var bottomPadding = 5.0

  var rightPadding = 5.0

  var isInvalid = false
    private set

  abstract fun asNode(id: Identifiable): DrawableProcessNode?

  override abstract fun builder(): Builder

  /**
   * Normalize the process model. By default this may do nothing.
   * @return The model (this).
   */
  fun normalize(): DrawableProcessModel? {
    return builder().apply { normalize(false) }.build().asM
  }

  open fun setNodes(nodes: Collection<DrawableProcessNode>) {
    super.setModelNodes(IdentifyableSet.processNodeSet(nodes))
    invalidate()
  }

  var vertSeparation: Double
    get() = layoutAlgorithm.vertSeparation
    set(vertSeparation) {
      if (layoutAlgorithm.vertSeparation != vertSeparation) {
        invalidate()
      }
      layoutAlgorithm.vertSeparation = vertSeparation
    }


  var horizSeparation: Double
    get() = layoutAlgorithm.horizSeparation
    set(horizSeparation) {
      if (layoutAlgorithm.horizSeparation != horizSeparation) {
        invalidate()
      }
      layoutAlgorithm.horizSeparation = horizSeparation
    }

  var defaultNodeWidth: Double
    get() = layoutAlgorithm.defaultNodeWidth
    set(defaultNodeWidth) {
      if (layoutAlgorithm.defaultNodeWidth != defaultNodeWidth) {
        invalidate()
      }
      layoutAlgorithm.defaultNodeWidth = defaultNodeWidth
    }


  var defaultNodeHeight: Double
    get() = layoutAlgorithm.defaultNodeHeight
    set(defaultNodeHeight) {
      if (layoutAlgorithm.defaultNodeHeight != defaultNodeHeight) {
        invalidate()
      }
      layoutAlgorithm.defaultNodeHeight = defaultNodeHeight
    }


  open fun invalidate() {
    isInvalid = true
  }

  fun resetLayout() {
    for (n in modelNodes) {
      n.setX(Double.NaN)
      n.setY(Double.NaN)
    }
    invalidate()
  }

  val endNodeCount: Int
    get() {
      var i = 0
      for (node in modelNodes) {
        if (node is EndNode<*, *>) {
          ++i
        }
      }
      return i
    }

  override fun getRef(): IProcessModelRef<DrawableProcessNode, DrawableProcessModel?, out @JvmWildcard RootClientProcessModel> {
    throw UnsupportedOperationException("Not implemented")
  }

  override fun getHandle(): Handle<out @JvmWildcard RootClientProcessModel> {
    return Handles.handle(handleValue)
  }

  fun getNode(nodeId: String): DrawableProcessNode? {
    for (n in modelNodes) {
      if (nodeId == n.id) {
        return n
      }
    }
    return null
  }

  fun setOwner(owner: String) {
    this.owner = SimplePrincipal(owner)
  }

  val startNodes: Collection<ClientStartNode>
    get() {
      val result = ArrayList<ClientStartNode>()
      for (n in modelNodes) {
        if (n is ClientStartNode) {
          result.add(n as ClientStartNode)
        }
      }
      return result
    }

  override fun addNode(node: DrawableProcessNode): Boolean {
    if (super.addNode(node)) {
      node.setOwnerModel(asM())
      // Make sure that children can know of the change.
      notifyNodeChanged(node)
      return true
    }
    return false
  }

  @Deprecated("Unsafe")
  override fun setNode(pos: Int, newValue: DrawableProcessNode): DrawableProcessNode {
    val oldValue = setNode(pos, newValue)

    newValue.setOwnerModel(asM())
    oldValue.setSuccessors(emptySet<Identified>())
    oldValue.setPredecessors(emptySet<Identified>())
    // TODO this is fundamentally unsafe, but we should get rid of [ClientProcessModel] anyway
    (oldValue as ClientProcessNode).setOwnerModel(null)

    for (pred in newValue.predecessors) {
      getNode(pred)!!.addSuccessor(newValue.identifier!!)
    }
    for (suc in newValue.successors) {
      getNode(suc)!!.addPredecessor(newValue.identifier!!)
    }

    return oldValue
  }

  override fun removeNode(nodePos: Int): DrawableProcessNode {
    val node = super.removeNode(nodePos)
    disconnectNode(node)
    return node
  }

  override fun removeNode(node: DrawableProcessNode): Boolean {
    if (node == null) {
      return false
    }
    if (super.removeNode(node)) {
      disconnectNode(node)
      return true
    }
    return false
  }

  private fun disconnectNode(node: DrawableProcessNode) {
    node.setPredecessors(emptyList<Identified>())
    node.setSuccessors(emptyList<Identified>())
    notifyNodeChanged(node)
  }

  open fun layout() {
    val diagramNodes = toDiagramNodes(modelNodes)
    if (layoutAlgorithm.layout(diagramNodes)) {
      var maxX = java.lang.Double.MIN_VALUE
      var maxY = java.lang.Double.MIN_VALUE
      for (n in diagramNodes) {
        n.target.setX(n.x + leftPadding)
        n.target.setY(n.y + topPadding)
        maxX = Math.max(n.right, maxX)
        maxY = Math.max(n.bottom, maxY)
      }
    }
  }

  private fun toDiagramNodes(modelNodes: Collection<DrawableProcessNode>): List<DiagramNode<DrawableProcessNode>> {
    val nodeMap = HashMap<Identified, DiagramNode<DrawableProcessNode>>()
    val result = modelNodes.map { node  ->
      DiagramNode(node).apply { node.identifier?.let { nodeMap[it] = this } ?: Unit }
    }

    for (diagramNode in result) {
      val modelNode = diagramNode.target
      modelNode.successors.asSequence()
          .map { nodeMap[it] }
          .filterNotNullTo(diagramNode.rightNodes)

      modelNode.predecessors.asSequence()
          .map { nodeMap[it] }
          .filterNotNullTo(diagramNode.leftNodes)
    }
    return result
  }

  companion object {

    const val NS_JBI = "http://adaptivity.nl/ProcessEngine/activity"

    const val NS_UMH = "http://adaptivity.nl/userMessageHandler"

    const val NS_PM = "http://adaptivity.nl/ProcessEngine/"

    internal const val PROCESSMODEL_NS = NS_PM
  }

}
